### ISSUE TITLE
Breaking unit-tests on AMD GPU

### ISSUE THREAD

**Issue by stratika** [beehive-lab/TornadoVM#71]
**Describe the bug**
There are 15 unit-tests in the `develop` branch failing when AMD is used as the default device.

**How To Reproduce**
In my setup the AMD is device `0:2`. I use the thunder machine which has an AMD gfx900 GPU.

```bash
git checkout develop
make
tornado-test.py --ea --verbose -J"-Dtornado.unittests.device=0:2"
```

**List of failing unit-tests**
``` bash
tornado -ea -Dtornado.unittests.verbose=True -Xmx6g -Dtornado.recover.bailout=False -Dtornado.unittests.device=0:2 uk.ac.manchester.tornado.unittests.tools.TornadoTestRunner  uk.ac.manchester.tornado.unittests.functional.TestLambdas

tornado -ea -Dtornado.unittests.verbose=True -Xmx6g -Dtornado.recover.bailout=False -Dtornado.unittests.device=0:2 uk.ac.manchester.tornado.unittests.tools.TornadoTestRunner  uk.ac.manchester.tornado.unittests.vectortypes.TestFloats

tornado -ea -Dtornado.unittests.verbose=True --debug -Xmx6g -Dtornado.recover.bailout=False -Dtornado.unittests.device=0:2 uk.ac.manchester.tornado.unittests.tools.TornadoTestRunner  uk.ac.manchester.tornado.unittests.reductions.TestReductionsFloats#testComputePi

tornado -ea -Dtornado.unittests.verbose=True -Xmx6g -Dtornado.recover.bailout=False -Dtornado.unittests.device=0:2 uk.ac.manchester.tornado.unittests.tools.TornadoTestRunner  uk.ac.manchester.tornado.unittests.profiler.TestProfiler

tornado -ea -Dtornado.unittests.verbose=True -Xmx6g -Dtornado.recover.bailout=False -Dtornado.unittests.device=0:2 uk.ac.manchester.tornado.unittests.tools.TornadoTestRunner  uk.ac.manchester.tornado.unittests.grid.TestGrid

tornado -ea -Dtornado.unittests.verbose=True -Xmx6g -Dtornado.recover.bailout=False -Dtornado.unittests.device=0:2 -Dtornado.device.desc=/home/thanos/repositories/tornadoVM/bin/sdk/examples/virtual-device-GPU.json -Dtornado.print.kernel=True -Dtornado.virtual.device=True -Dtornado.print.kernel.dir=/home/thanos/repositories/tornadoVM/bin/sdk/virtualKernelOut.out uk.ac.manchester.tornado.unittests.tools.TornadoTestRunner  uk.ac.manchester.tornado.unittests.virtual.TestVirtualDeviceKernel#testVirtualDeviceKernelGPU

tornado -ea -Dtornado.unittests.verbose=True -Xmx6g -Dtornado.recover.bailout=False -Dtornado.unittests.device=0:2 -Dtornado.device.desc=/home/thanos/repositories/tornadoVM/bin/sdk/examples/virtual-device-CPU.json -Dtornado.print.kernel=True -Dtornado.virtual.device=True -Dtornado.print.kernel.dir=/home/thanos/repositories/tornadoVM/bin/sdk/virtualKernelOut.out uk.ac.manchester.tornado.unittests.tools.TornadoTestRunner  uk.ac.manchester.tornado.unittests.virtual.TestVirtualDeviceKernel#testVirtualDeviceKernelCPU

tornado -ea -Dtornado.unittests.verbose=True -Xmx6g -Dtornado.recover.bailout=False -Dtornado.unittests.device=0:2 -Dtornado.device.desc=/home/thanos/repositories/tornadoVM/bin/sdk/examples/virtual-device-GPU.json -Dtornado.virtual.device=True -Dtornado.feature.extraction=True -Dtornado.features.dump.dir=/home/thanos/repositories/tornadoVM/bin/sdk/virtualFeaturesOut.out uk.ac.manchester.tornado.unittests.tools.TornadoTestRunner  uk.ac.manchester.tornado.unittests.virtual.TestVirtualDeviceFeatureExtraction#testVirtualDeviceFeaturesGPU

tornado -ea -Dtornado.unittests.verbose=True -Xmx6g -Dtornado.recover.bailout=False -Dtornado.unittests.device=0:2 -Dtornado.device.desc=/home/thanos/repositories/tornadoVM/bin/sdk/examples/virtual-device-CPU.json -Dtornado.virtual.device=True -Dtornado.feature.extraction=True -Dtornado.features.dump.dir=/home/thanos/repositories/tornadoVM/bin/sdk/virtualFeaturesOut.out uk.ac.manchester.tornado.unittests.tools.TornadoTestRunner  uk.ac.manchester.tornado.unittests.virtual.TestVirtualDeviceFeatureExtraction#testVirtualDeviceFeaturesCPU
```

Resulting in:
```
==================================================
              Unit tests report 
==================================================

{'[FAILED]': 15, '[PASS]': 321}
Coverage: 95.54%

==================================================
```

**Comment by jjfumero** [comment]
- ComputePi is expected to fail (due to barriers on AMD). 
-  @gigiblender , can you take a look at the VirtualDevice tests on AMD?

**Comment by stratika** [comment]
I fixed the `TestGrid`. And I had a look in the `TestVirtualLayer`. It seems that there is a method that checks if there are enough device (`enoughdevices()`). This method invokes another method called `before()`, which attempts to swap two devices with a new index. However, the array for the `flatBackends` has length 1, so it produces an `ArrayOutOfBound Exception`.

@gigiblender if you can, please have a look. I cannot understand how it should work and why it sees only 1 device.

**Comment by gigiblender** [comment]
@stratika @jjfumero I will have a look in the following days

**Comment by vsilaev** [comment]
The issue with profiler is that 
```json
        "COPY_OUT_TIME": "0",
        "COPY_IN_TIME": "0",
```
It looks like in reality both values are `NULL` and we got NPE in  assert statements

**Comment by stratika** [comment]
> The issue with profiler is that
> 
> ```json
>         "COPY_OUT_TIME": "0",
>         "COPY_IN_TIME": "0",
> ```
> 
> It looks like in reality both values are `NULL` and we got NPE in assert statements

You are correct, it seems that the AMD drivers return `0` for the events related with these two operations.

**Comment by vsilaev** [comment]
There is an issue with TestBatches as well -- when testing on devices that has < 2GB ram some test fails with pretty clear description about insufficient memory.

**Comment by stratika** [comment]
> There is an issue with TestBatches as well -- when testing on devices that has < 2GB ram some test fails with pretty clear description about insufficient memory.

By "devices", do you refer to AMD GPUs with memory capacity < 2GB? If yes, the AMD GPU in our server has 4GB, so it may be the case that we have not seen this issue yet. Can you please open a new issue and provide the description and test case.

**Comment by jjfumero** [comment]
> The issue with profiler is that
> 
> ```json
>         "COPY_OUT_TIME": "0",
>         "COPY_IN_TIME": "0",
> ```
> 
> It looks like in reality both values are `NULL` and we got NPE in assert statements

This is related to the AMD driver:

See the related issue here: https://github.com/aparapi/aparapi/issues/24

**Comment by vsilaev** [comment]
@jjfumero then it strange why profiler test reports about some null value in assert statement.

**Comment by jjfumero** [comment]
True, we will look into that. I think @stratika has a PR almost ready.

**Comment by gigiblender** [comment]
#75 should fix the tests for virtual devices

**Comment by stratika** [comment]
Here is an update, regarding this issue:

Some unit-tests are failing in our `AMD Radeon RX Vega gfx900 GPU`, with driver `2766.4`.
My guess is that our driver (which is old) has some issues with the following tests, since these tests are passing on other AMD GPUs with latest drivers. This is the list of the test and the reason of failure.
- `TestLambdas`: These test are failing with a -46 `CL_INVALID_KERNEL_NAME` error code. I experimented and concluded that the `$` character in the kernel name is triggering this failure, despite being a valid character for kernel names.
- `TestFloats`: In these set of unit-tests the only failing test is the `testSimpleDotProductFloat8`. The behaviour is odd, as the kernel is functional on other devices and the logic seems correct. However, it returns a wrong result. In this case also, latest driver seems to pass this test.
- `TestReductionsFloats#testComputePi` is expected to fail (due to barriers on AMD), as mentioned earlier by @jjfumero.

Now, the following unit-tests are passing with the patch in PR #76 .
- `TestProfiler` was failing because the value that is returned from the events  `CL_PROFILING_COMMAND_END`, `CL_PROFILING_COMMAND_START` is zero. 
These two flags are used by the OpenCL standard to return **[a 64-bit value that describes the current device time counter in nanoseconds](https://www.khronos.org/registry/OpenCL/sdk/1.0/docs/man/xhtml/clGetEventProfilingInfo.html)**. This was triggering the tests to fail.
From our understanding, when we execute on the AMD GPU of our server, the reporting `TOTAL_KERNEL_TIME` from the TornadoVM profiler includes also the copy_in and copy_out times for the data transfers. And the AMD driver (2766.4) returns zero for the events.
- `TestGrid` was failing because the `AMDScheduler` class was not updated to support the Grid and GridScheduler features.

And the `TestVirtualDeviceKernel` and `TestVirtualDeviceFeatureExtraction` unit-tests are fixed in PR #75 by @gigiblender.

**Comment by jjfumero** [comment]
With this PR https://github.com/beehive-lab/TornadoVM/pull/91 we have 100% tests passing on AMD and ROCm drivers.
